pcdsdevices/slits.py

```diff
@@ -0,0 +1,321 @@
+axes into a Slits object. This allows an operator to manipulate the center and
+width in two dimensions of a small aperture. The classes below allow both
+individual parameters of the aperture and the Slit as a whole to be controlled
+and scanned. The :class:`.Slits` instantiates four sub-devices `xwidth`,
+`xcenter`, `ycenter`, `ywidth`. These are each represented by
+:class:`.SlitPositioner`. The main :class:`.Slits` class assumes that most of
+the manipulation will be done on the size of the aperture not the position,
+however, if control of the center is desired the `center` sub-devices can be
+used.
+"""
+import logging
+
+import numpy as np
+from ophyd.status import wait as status_wait
+from ophyd.pv_positioner import PVPositioner
+from ophyd import (Device, EpicsSignal, EpicsSignalRO, Component as C,
+                   FormattedComponent as FC)
+
+logger = logging.getLogger(__name__)
+
+
+class SlitPositioner(PVPositioner, Device):
+    """
+    Abstraction of Slit Axis
+
+    Each adjustable parameter of the slit (center, width) can be modeled as a
+    motor in itself, even though each controls two different actual motors in
+    reality, this gives a convienent interface for adjusting the aperture size
+    and location with out backwards calculating motor positions
+
+    Parameters
+    ----------
+    prefix : str
+        The prefix location of the slits, i.e MFX:DG2
+
+    slit_type : XWIDTH, YWIDTH, XCENTER, YCENTER
+        The aspect of the slit position you would like to control with this
+        specific motor
+
+    name : str
+        Alias for the axis
+
+    limits : tuple, optional
+        Limits on the motion of the positioner. By default, the limits on the
+        setpoint PV are used if None is given.
+
+    See Also
+    --------
+    ophyd.PVPositioner
+        SlitPositioner inherits directly from PVPositioner.
+    """
+    setpoint = FC(EpicsSignal, "{self.prefix}:{self._dirshort}_REQ")
+    readback = FC(EpicsSignalRO, "{self.prefix}:ACTUAL_{self._dirlong}")
+    done = C(EpicsSignalRO, ":DMOV")
+
+    _default_read_attrs = ['readback']
+
+    def __init__(self, prefix, *, slit_type="", name=None,
+                 limits=None, **kwargs):
+        # Private PV names to deal with complex naming schema
+        self._dirlong = slit_type
+        self._dirshort = slit_type[:4]
+        # Initalize PVPositioner
+        super().__init__(prefix, limits=limits, name=name, **kwargs)
+
+    @property
+    def egu(self):
+        """Engineering units"""
+        return self._egu or self.readback._read_pv.units
+
+    def _setup_move(self, position):
+        # This is subclassed because we need `wait` to be set to False unlike
+        # the default PVPositioner method. `wait` set to True will not return
+        # until the move has completed
+        logger.debug('%s.setpoint = %s', self.name, position)
+        self.setpoint.put(position, wait=False)
```

Comment:
It bugs me that we have to do this. There's some weird stuff going on in the slits IOC.

Reply:
I'll add it to the list of things to modify if we ever tweak this IOC